exchange/currency.go

```diff
@@ -0,0 +1,222 @@
+package exchange
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/buger/jsonparser"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+)
+
+const latestConversionRatesUrl = "http://currency.prebid.org/latest.json"
+const defaultCurrencyConversionError = "No currency conversion available."
+
+type currencyData struct {
+	DataAsOfDate string      `json:"dataAsOf"`
+	Conversions  conversions `json:"conversions"`
+}
+
+type conversions map[string]rates
+
+type rates map[string]float64
+
+func getCurrency(seatBids *[]openrtb.SeatBid, bidRequest *openrtb.BidRequest, adapterExtra *map[openrtb_ext.BidderName]*seatResponseExtra) string {
+	currencyToUse := defaultAdServerCurrency
+	var requestExt openrtb_ext.ExtRequest
+
+	if bidRequest.Ext != nil {
+		if err := json.Unmarshal(bidRequest.Ext, &requestExt); err != nil {
+			// Error unmarshalling bidRequest.Ext. Treat as if currency was not set in request.
+			return currencyToUse
+		}
+
+		requestCurrency := strings.ToUpper(requestExt.Currency.AdServerCurrency)
+		// First check if currency was set in request
+		if requestCurrency != "" {
+			newSeatBids := make([]openrtb.SeatBid, 0, len(*seatBids))
+
+			// Iterate through bidder currencies and make sure they match with request currency
+			for _, seatBid := range *seatBids {
+				var bidderErrs []interface{}
+				bidderName, ok := openrtb_ext.GetBidderName(seatBid.Seat)
```

Comment:
note: not 100% sure if this is the best way to get the bidder from the seatBid object. if this is not reliable please let me know so I search for alternatives.

Reply:
This works.

We'll have to revisit this, along with many other parts of the auction logic, for #271... but I don't think you should worry about that now.